### PR TITLE
feat(activity-logging): Adds activity logging to insights

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

Building on #8931 for #8545 

## Changes

This adds activity logging for Insights. Since the surface area of possible changes is large across insights and might cause errors from changes that can't be serialized. It also changes activity logging so that it logs and swallows exceptions caused when trying to log activity

Screenshot 2022-03-20 at 14 34 53

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

adds API tests, and created, changed, and deleted insights locally and observed they were saved to the DB

### CODE REVIEW

**Comment by mariusandra** [comment]
Hey, love the initiative!

Before merging this though, we might want a definite answer on what's going to happen to the dashboard and insight many-to-many task. I think we should just do it, that is to say separate `insights` from `dashboard_items` and be done with it. Then we can both add other things than insights onto dashboards, and make a choice on how many dashboards an insights belongs to on our terms. 

I see us creating a new table `insights` for insights, and migrating all of them over from the existing `dashboarditem` table. It may be annoying to edit a lot of log entries in a migration if we make such a big shift in the data, thus we might want to keep this open for a little bit... maybe.

**Comment by pauldambra** [comment]
> It may be annoying to edit a lot of log entries in a migration if we make such a big shift in the data

The fun thing about this is that we have separated storage of the log and storage of the model, so we have several ways we can handle changes to models.

If it's "just" a storage move we might not even care. The code that logs the activity would still log it.

Otherwise

The first is to change the scope. If the model has changed significantly enough that the scope is different. 

`log_activity(scope="insight")` -> `log_activity(scope="new insights")` 

(NB we shouldn't use "new insights" in real code :))

The other is to add a `version` to the `detail` we log with the activity

`log_activity(scope="insight", detail=Detail(version=2))` 

Then the front end `describer` would be taught how to describe version 2

----

So I'd be tempted to add this (other review pending obvs). Better to have the data than not. 

(plus I'm sure we were talking about changing the insight/dashboard item relationship 6 months ago 🤣 )

**Comment by mariusandra** [comment]
Agreed that at the end of the day we don't really care what happens to the data at this state. I won't be a blocker, don't worry 😅 .

However I'm talking about something that we are going to do [sooner rather than later](https://github.com/posthog/product-internal/pull/290/#discussion_r830199634). We're at a place where we're making dumb product decisions because of this self-imposed technical constraint that we avoid fixing for no real reason... So I was literally planning on working on this next week or next sprint. 😄

**Comment by pauldambra** [comment]
Marked this as draft until after insights work is finished...

Will now nag @mariusandra periodically until it is done

**Comment by pauldambra** [comment]
@mariusandra since we know that #9416 won't (significantly) change the insight table. I think this is good to go

File: posthog/models/activity_logging/activity_log.py [hunk 1]
```diff
@@ -78,14 +81,34 @@ class Meta:
...
+        "created_at",
+        "refreshing",
+        "dive_dashboard",
+        "updated_at",
+        "type",
+        "funnel",
+        "deprecated_tags",
+        "last_modified_at",
+        "layouts",
+        "color",
```
Comment (mariusandra) [inline posthog/models/activity_logging/activity_log.py:99]: We can probably remove these two from here?

File: posthog/models/activity_logging/activity_log.py [hunk 2]
```diff
@@ -78,14 +81,34 @@ class Meta:
...
+        "order",
+        "result",
+        "dashboard",
```
Comment (mariusandra) [inline posthog/models/activity_logging/activity_log.py:102]: Should we track `dashboards` now?

**Review by mariusandra — CHANGES_REQUESTED** [review]
Seems to make sense, just let's already track the new fields. Other than the inline comments, good to go!

File: posthog/models/activity_logging/activity_log.py (hunk 2 above)
Comment (pauldambra) [reply posthog/models/activity_logging/activity_log.py:102]: To check on these two bits of review...

This is a list of things to exclude on the server. So it says "don't track these deprecated fields" 

Have you misunderstood the code or have I misunderstood the review 🤣 

(we do pick up dashboards field in changes)

File: posthog/api/insight.py [hunk 3]
```diff
@@ -172,9 +186,27 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Insight:
 
         # Manual tag creation since this create method doesn't call super()
         self._attempt_set_tags(tags, insight)
+
+        name_for_logged_activity = (
+            insight.name if insight.name is not None and len(insight.name) > 0 else insight.derived_name
+        )
```
Comment (mariusandra) [inline posthog/api/insight.py:192]: Nit: I'm not python enough to know if this is the best way to do this... but it looks long

File: posthog/models/activity_logging/activity_log.py (hunk 2 above)
Comment (mariusandra) [reply posthog/models/activity_logging/activity_log.py:102]: imagine I put a homer bush gif here

**Review by mariusandra — APPROVED** [review]

File: posthog/api/insight.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/api/insight.py:192]: If I'm understanding correctly:

`if insight.name is not None and len(insight.name) > 0` is same as `if insight.name` ? (`[]` and `''` and any other thing that should be falsy is falsy, unlike js) xD

Comment (pauldambra) [reply posthog/api/insight.py:192]: 🤯 

But programming languages are _supposed_ to be confusing! :)

**Pull request merged**